crates/tracing/src/stack.rs

```diff
@@ -0,0 +1,155 @@
+//! This module is centered around a thread-local Stack of Span references,
+//! with some public functions for pushing and accessing values from the stack.
+//! This is primarily used by the autotrace proc macro in holochain_tracing_macros,
+//! to automatically push new child spans onto the stack when entering a new function frame.
+//! Using functions like `with_top` allow the user to access the stack directly, for situations
+//! like needing to take a span context and send it into another thread, or out of the process entirely.
+
+use std::cell::RefCell;
+use std::rc::Rc;
+
+use crate::Span;
+
+/// This enum defines how to handle situations where we expect there to be a Span
+/// on the stack, but there is none.
+#[allow(dead_code)]
+enum Mode {
+    /// Panic when finding an empty stack. Useful for quickly discovering gaps in tracing coverage
+    Panic,
+    /// Emit a warning and a full backtrace. Note, this is very noisy and slow!
+    Backtrace,
+    /// Ignore cases of an empty stack, and just return a null span.
+    Noop,
+}
+
+const MODE: Mode = Mode::Noop;
+
+thread_local! {
+    static SPANSTACK: RefCell<SpanStack> = RefCell::new(SpanStack::default());
+}
+
+lazy_static! {
+    static ref NOOP_SPAN: Span = Span::noop();
+}
+
+#[derive(Default)]
+struct SpanStack(Vec<Rc<Span>>);
+
+impl SpanStack {
+    // fn len(&self) -> usize {
+    //     self.0.len()
+    // }
+
+    fn push_span(&mut self, span: Rc<Span>) {
+        self.0.push(span);
+    }
+
+    // fn push_fn<F: FnOnce(&Span) -> Rc<Span>>(&mut self, f: F) {
+    //     if let Some(top) = self.0.last() {
+    //         let successor = f(top);
+    //         self.0.push(successor);
+    //     } else {
+    //         warn!("Using push_fn, but the stack is empty!");
+    //     }
+    // }
+
+    fn pop(&mut self) {
+        let _ = self.0.pop();
+    }
+
+    fn top(&self) -> Option<&Span> {
+        self.0.last().map(|s| (*s).as_ref())
+    }
+
+    // fn is_empty(&self) -> bool {
+    //     self.0.is_empty()
+    // }
+}
+
+/// A guard to track the lifetime of an item on the stack. Items are popped from the stack
+/// when these guards are dropped.
+/// Each guard corresponds to a single item in the stack, and contains Rc references
+/// to each item below this item in the stack.
+/// This is such that if a guard for an item is dropped, the item will not be popped from the stack
+/// if there are still guards from higher-up items still on the stack.
+pub struct SpanStackGuard {
+    _spans: Vec<Rc<Span>>,
+}
+
+impl SpanStackGuard {
+    pub fn new(span: Span) -> Self {
+        let span = Rc::new(span);
+        let _spans = SPANSTACK.with(|stack| {
+            let mut stack = stack.borrow_mut();
+            stack.push_span(span.clone());
+            stack.0.clone()
```

Comment:
I wonder if you should consider [shrinkwraping](https://docs.rs/crate/shrinkwraprs/0.2.3) the stack to simplify some of this code and treat the guard as vector. No need for `push_span` method, for example, or to access the `0` element of the struct.

Reply:
In retrospect it may not work due to the fact that push_span is more specialized than a regular vector push. Nevermind.